### PR TITLE
Amazon VOD conversion to modular, and ReuseInvoker

### LINKED ISSUE #143: Fixes and enhancements for the ReusePythonInvoker
@peak3d was gracious enough to submit #142, fast fixes to support ReuseInvoker. I temporarily merged into [enh-reuseinvoker](https://github.com/Sandmann79/xbmc/tree/enh-reuseinvoker), where we can test and further improve it.

@Sandmann79 it requires testing on your part, in the meanwhile I'll propagate the changes to AmazonVOD.

### PR DESCRIPTION
**Pull request by Varstahl**
Fixes #143

There is quite a list of changes with this one. Before going in the detail, @Sandmann79 I need quite some in-depth review/test for several features I wasn't able to test. I went as far as I could with static and dynamic debugging, but there are things like tvdb integration, captcha and amazon.tld playback that I couldn't test. It should work, but prepare for the worst.

## Changes

- Converted the single `default.py` into several different modules. It could be still extended further to provide better backend logic and possibly performance, but atm it should just work.
- Two singleton classes, `Globals` and `Settings`, are the most significant change:
  - **Globals**; contains all the previously global variables and initialisation that were scattered all across the place. If it's a constant or should be consider global, it now sits there. As a rationale, I've kept the constants as attributes, and all the dynamics as non-existent attributes, fetched through `__getattr__` to prevent unexpected manipulations. For the same reason I (temporarily?) disabled `__setattr__` and `__delattr__`,
  - **Settings**: another singleton, fairly similar to Globals in usage, but way dynamic in nature. It doesn't store nor cache anything, which should avoid problems with the variables access/caching. We don't currently have a service so it's probably not necessary, but a bit of future proofing can't hurt.
- The other two singletons, `AmazonTLD` and `PrimeVideo`, are more or less wrappers for functions previously available in the global namespace. After initialisation, they can be accessed through `g.amz` and `g.pv`. A lot of platform specific initialisation related to amazon.tld/primevideo.com, previously sitting inside `default.py`, has been moved into their `__init__`. In the future it should be possible to further abstract into a parent class both AmazonTLD and PrimeVideo, to better synchronise features such as the BrowseRoot, the export of the library, or the Watchlist settings.
  - **PrimeVideo**: pretty much the same, since it was already self contained the migration was quite painless.
  - **AmazonTLD**: I've slightly extended the functions, adding a couple hooks such as `_initialiseDB` and `BrowseRoot`.
- The logging module features a slightly modified `Log` function, with embedded log level. It's now possible to call it like this `Log('message', Log.INFO)`. This removes the need to import `xbmc` everywhere, and allows eventual level remapping in the future without headaches.
- I removed a lot of the `sys.argv` calls. Most were references to the plugin name to be used in the URI. They have been replaced with a fetch to the global instantiation, now sitting under `g.pluginid`.
- The global `platform` **is no longer decimal**. I converted it into a binary encoded mask, removed the LibreElec variable from the local/global namespace, and included it in the platform variable itself. With this linearisation we can not only track the platform, but even the variant if need be (just like with LE), and in the same place. [Example here](https://github.com/Sandmann79/xbmc/commit/fab71e6296a43250f7098e5f06ace82129b74ac3#diff-61ba02e2d92b786cc74537fe5b16294bR826).
- Prepended an underscore to functions (and classes) that didn't need to be treated as public APIs, and nested them where possible. One such case is inside `PlayVideo`, where a lot of the subroutines used a lot of variables in the global namespace. Since there was no harm nor foul, I nested them all inside PlayVideo to reduce the headaches, especially considering further changes were not needed.
- `AgeSettings`, is now renamed `_AgeSettings` and lives inside the global `AgeRestrictions` class. Given the nature and scope, I decided to wrap all the age related code inside AgeRestrictions. It is currently not a singleton and doesn't cache anything, in spirit with `Settings`.
- `tmdb` and `tvdb` are no longer obtained through base64 decoding, and are instead embedded as binary string constants. If there was a good reason why this wasn't previously done (different endians?), we might want to revert that.
- **Put a dam around the free [exec()](https://github.com/Varstahl/xbmc/commit/d3a0646e051df3e8ab78eed22d6e485e43f92935#diff-b7c75170872617b4686f3dd0f213e83eR91), but it won't hold. We need to get rid of it.**

## Conclusions

These should be all the major changes in this PR. I've applied PEP8 with a few exceptions here and there to improve readability. I won't merge immediately so that we can talk about it a bit, and see if we want to revert some parts of it, or further abstract and enhance it.

When it came to your code for Amazon.tlds, I tried to touch it as little as possible given that I couldn't fully test it. I saw some potential for slimming and linearisation, but I'll leave you to it, you certainly know that code more than me.

So… this is one hell of a change, and I know I took a lot of concepts that aren't properly coming from the Python philosophy (and generally never cared much for it), but I'm open to all kinds of feedback.

### CODE REVIEW

File: plugin.video.amazon-test/resources/lib/startup.py [hunk 1]
```diff
@@ -84,8 +83,8 @@ def EntryPoint():
         g.pv.Browse(None if 'path' not in args else args['path'])
     elif mode == 'PrimeVideo_Search':
         g.pv.Search()
-    elif None is not re.match(r'^[a-zA-Z]+$', mode):
+    elif re.match(r'^[a-zA-Z]+$', mode):
         cmd = mode + '()'
```
Comment (Varstahl) [inline plugin.video.amazon-test/resources/lib/startup.py]: This is still unsafe and exploitable. Are there unmapped method calls we need to make explicit, or can we get rid of this now?

File: plugin.video.amazon-test/resources/lib/startup.py [hunk 2]
```diff
@@ -84,8 +83,8 @@ def EntryPoint():
...
-        # if mode in g.amz:
-        #     cmd = 'g.amz.' + cmd
+        if mode in dir(g.amz):
+            cmd = 'g.amz.' + cmd
```
Comment (Varstahl) [inline plugin.video.amazon-test/resources/lib/startup.py]: I believe you forgot to add the brackets (`cmd = 'g.amz.' + cmd + '()'`). Unless I'm misreading this, as it stands it won't call the method.

File: plugin.video.amazon-test/resources/lib/startup.py (hunk 1 above)
Comment (Sandmann79) [reply plugin.video.amazon-test/resources/lib/startup.py]: Oh just forget to add it. We can use this, to detect existing methods:
`elif mode in dir(g.amz) + globals().keys():`
That should be safer, then the old call.

Comment (Varstahl) [reply plugin.video.amazon-test/resources/lib/startup.py]: Done. I'm still not aware of any raw calls like that though, so for the moment I left the list empty.

Comment (Sandmann79) [reply plugin.video.amazon-test/resources/lib/startup.py]: I had to changed this, cause of compatibility to Kodi 17. The `dir` call doesn't work as expected there. I hope, i didn't forget any method.

**Comment by Varstahl** [comment]
There are a few things I wanted to tweak and re-check, like cbec386b943ded57ef35be0da6eb28153be0a1c3, but haven't found the time yet. But I agree, there aren't many loose ends left, if any.

**Comment by Sandmann79** [comment]
That's doesn't work very well. The problem here is to get the DBID from Kodi and that's doesn't work at home screen widgets. I think it's better to use an own database, where this resume points are stored and ignoring Kodi ones.

**Comment by Varstahl** [comment]
I guess that's not too complicated. It's doable and the resume-point can be reintroduced with a ListItem option if I'm not wrong, so that's no biggie. I was mostly considering the analysis of the PrimeVideo.com endpoint though, since the tracking doesn't seem to work at all atm.

**Pull request merged**